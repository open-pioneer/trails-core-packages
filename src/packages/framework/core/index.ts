// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
export { Error, getErrorChain, isAbortError, throwAbortError, createAbortError } from "./error";
export { EventEmitter, type EventSource, type EventNames } from "./events";
export { destroyResource, type Resource } from "./resources";
export { createLogger, type Logger, type LogLevel, type LogMethod } from "./Logger";
export { createManualPromise, type ManualPromise } from "./utils";
