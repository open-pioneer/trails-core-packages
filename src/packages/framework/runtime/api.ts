// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0

/* eslint-disable @typescript-eslint/no-explicit-any */
export type ApiMethod = (...args: any[]) => any;

/**
 * A record of exposed functions.
 */
export type ApiMethods = Record<string, ApiMethod>;

/**
 * Provides a set of methods to the application's web component API.
 */
export interface ApiExtension {
    /**
     * Returns a set of methods that will be added to the web component's API.
     */
    getApiMethods(): Promise<ApiMethods>;
}

/**
 * A service provided by the system.
 * Used by the runtime to assemble the public facing API.
 */
export interface ApiService {
    /**
     * Called by the runtime to gather methods that should be available from the web component's API.
     */
    getApi(): Promise<ApiMethods>;
}

/**
 * A service provided by the system, useful for accessing values that are global to the application.
 */
export interface ApplicationContext {
    /**
     * The web component's host element.
     * This dom node can be accessed by the host site.
     */
    getHostElement(): HTMLElement;

    /**
     * The current web component's shadow root.
     */
    getShadowRoot(): ShadowRoot;

    /**
     * The node containing the rest of the application _inside_ the current web component's shadow dom.
     */
    getApplicationContainer(): HTMLElement;

    /**
     * Returns the current locale of the application.
     *
     * E.g. "de-DE"
     */
    getLocale(): string;

    /**
     * Returns the locales supported by the application.
     *
     * E.g. ["de", "en"]
     */
    getSupportedLocales(): readonly string[];
}

declare module "./ServiceRegistry" {
    interface ServiceRegistry {
        "runtime.ApiService": ApiService;
        "runtime.ApplicationContext": ApplicationContext;
    }
}
