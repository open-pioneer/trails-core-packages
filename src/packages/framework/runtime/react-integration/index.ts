export {
    useServiceInternal,
    useServicesInternal,
    usePropertiesInternal,
    useIntlInternal,
    type UseServiceOptions
} from "./hooks";

/* Re-exported for test-support. Do not use directly. */
export { PackageContext, type PackageContextMethods } from "./PackageContext";
