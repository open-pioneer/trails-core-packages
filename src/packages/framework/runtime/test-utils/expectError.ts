export function expectError(impl: () => unknown): Error {
    try {
        impl();
        throw new Error("expected error!");
    } catch (e) {
        if (e instanceof Error) {
            return e;
        }
        throw new Error("unexpected error value, not an instance of Error");
    }
}

export function expectAsyncError(impl: () => Promise<unknown>): Promise<Error> {
    const promise = impl();
    return promise.then(
        () => {
            throw new Error("expected error!");
        },
        (e) => {
            if (e instanceof Error) {
                return e;
            }
            throw new Error("unexpected error value, not an instance of Error");
        }
    );
}
