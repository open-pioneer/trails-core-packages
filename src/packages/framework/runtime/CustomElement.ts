// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { ComponentType } from "react";
import {
    createAbortError,
    createLogger,
    createManualPromise,
    destroyResource,
    Error,
    isAbortError,
    ManualPromise,
    Resource,
    throwAbortError
} from "@open-pioneer/core";
import { ErrorId } from "./errors";
import { ApplicationMetadata, ObservableBox, PackageMetadata } from "./metadata";
import { PackageRepr, createPackages } from "./service-layer/PackageRepr";
import { ServiceLayer } from "./service-layer/ServiceLayer";
import { getErrorChain } from "@open-pioneer/core";
import { ReactIntegration } from "./react-integration/ReactIntegration";
import { ApiMethods, ApiService } from "./api";
import { createBuiltinPackage, RUNTIME_API_SERVICE } from "./builtin-services";
import { ReferenceSpec } from "./service-layer/InterfaceSpec";
import { PropertiesRegistry } from "./PropertiesRegistry";
import { AppI18n, initI18n } from "./i18n";
const LOG = createLogger("runtime:CustomElement");

/**
 * Options for the {@link createCustomElement} function.
 */
export interface CustomElementOptions {
    /**
     * Rendered UI component.
     */
    component?: ComponentType<Record<string, string>>;

    /**
     * Application metadata (packages, services etc.).
     * This is usually autogenerated by importing the virtual `"open-pioneer:app"` module.
     */
    appMetadata?: ApplicationMetadata;

    /**
     * Application defined configuration.
     * All instances of the web component will share this static configuration.
     *
     * These will override default properties of the application's packages.
     */
    config?: ApplicationConfig;

    /**
     * Function to provide additional application defined configuration parameters.
     *
     * Compared to {@link config}, this function receives a context object
     * that allows the developer to provide dynamic properties on a per-application basis.
     *
     * Parameters returned by this function take precedence over the ones defined by {@link config}.
     */
    resolveConfig?(ctx: ConfigContext): Promise<ApplicationConfig | undefined>;

    /**
     * Whether the shadow root element is accessible from the outside.
     * Defaults to `false` in production mode and `true` during development to make testing easier.
     */
    openShadowRoot?: boolean;
}

/**
 * A context object that is passed to the `resolveProperties` function.
 */
export interface ConfigContext {
    /**
     * Returns an attribute from the application's root node.
     */
    getAttribute(name: string): string | undefined;
}

/**
 * Runtime application configuration.
 */
export interface ApplicationConfig {
    /**
     * Set this value to a locale string (e.g. "en") to for the application's locale.
     * The default behavior is to choose an appropriate locale for the current user based
     * on the browser's settings.
     *
     * The locale must be supported by the application.
     */
    locale?: string | undefined;

    /**
     * Properties specified here will override default properties of the application's packages.
     */
    properties?: ApplicationProperties | undefined;
}

/**
 * Allows the application to override default properties in all packages.
 */
export interface RawApplicationProperties {
    /**
     * Key: the name of the package.
     * Value: A record of configuration properties (key/value pairs).
     *
     * Properties will override default property values in the package.
     */
    [packageName: string]: Record<string, unknown>;
}

/**
 * Allows the application to override default properties in all packages.
 *
 * Properties are typed when the package contains type definitions for them.
 */
export type ApplicationProperties = RawApplicationProperties & Partial<PropertiesRegistry>;

/**
 * The interface implemented by web components produced via {@link createCustomElement}.
 */
export interface ApplicationElement extends HTMLElement {
    /** Resolves to the element's API when the application has started. */
    when(): Promise<ApiMethods>;
}

/**
 * The class returned by a call to {@link createCustomElement}.
 */
export interface ApplicationElementConstructor {
    new (): ApplicationElement;
}

/**
 * Creates a new custom element class (web component) that can be registered within a DOM.
 *
 * @example
 * ```ts
 * import * as appMetadata from "open-pioneer:app";
 *
 * const CustomElementClazz = createCustomElement({
 *   component: <div>Hello World!</div>,
 *   appMetadata
 * });
 * customElements.define("sample-element", CustomElementClazz);
 * ```
 */
export function createCustomElement(options: CustomElementOptions): ApplicationElementConstructor {
    class PioneerApplication extends HTMLElement implements ApplicationElement {
        #shadowRoot: ShadowRoot;
        #state: ElementState | undefined;

        static get observedAttributes(): string[] {
            return [];
        }

        constructor() {
            super();

            const mode = options.openShadowRoot ?? import.meta.env.DEV ? "open" : "closed";
            this.#shadowRoot = this.attachShadow({
                mode: mode
            });
        }

        connectedCallback() {
            LOG.debug("Launching application");

            if (this.#state) {
                this.#state.destroy();
            }

            this.#state = new ElementState(this, this.#shadowRoot, options);
            this.#state.start();
        }

        disconnectedCallback() {
            LOG.debug("Shutting down application");

            this.#state?.destroy();
            this.#state = undefined;

            LOG.debug("Application destroyed");
        }

        when() {
            if (!this.#state) {
                return Promise.reject(
                    new Error(
                        ErrorId.NOT_MOUNTED,
                        "Cannot use the application's API because the HTML element has not yet been mounted into the DOM."
                    )
                );
            }

            return this.#state.whenAPI();
        }
    }
    return PioneerApplication;
}

class ElementState {
    private hostElement: HTMLElement;
    private shadowRoot: ShadowRoot;
    private options: CustomElementOptions;

    // Public API
    private apiPromise: ManualPromise<ApiMethods> | undefined; // Present when callers are waiting for the API
    private api: ApiMethods | undefined; // Present once started

    private state = "not-started" as "not-started" | "starting" | "started" | "destroyed";
    private locale: string | undefined;
    private container: HTMLDivElement | undefined;
    private config: ApplicationConfig | undefined;
    private serviceLayer: ServiceLayer | undefined;
    private reactIntegration: ReactIntegration | undefined;
    private stylesWatch: Resource | undefined;

    constructor(hostElement: HTMLElement, shadowRoot: ShadowRoot, options: CustomElementOptions) {
        this.hostElement = hostElement;
        this.shadowRoot = shadowRoot;
        this.options = options;
    }

    start() {
        if (this.state !== "not-started") {
            throw new Error(ErrorId.INTERNAL, `Cannot start element in state '${this.state}'`);
        }

        this.state = "starting";
        this.startImpl().catch((e) => {
            // TODO: Error splash?
            this.destroy();
            if (!isAbortError(e)) {
                logError(e);
            }
        });
    }

    destroy() {
        this.state = "destroyed";
        this.apiPromise?.reject(createAbortError());
        this.reactIntegration = destroyResource(this.reactIntegration);
        this.shadowRoot.replaceChildren();
        this.container = undefined;
        this.serviceLayer = destroyResource(this.serviceLayer);
        this.stylesWatch = destroyResource(this.stylesWatch);
    }

    whenAPI(): Promise<ApiMethods> {
        if (this.api) {
            return Promise.resolve(this.api);
        }

        const apiPromise = (this.apiPromise ??= createManualPromise());
        return apiPromise.promise;
    }

    private async startImpl() {
        const { options, shadowRoot, hostElement } = this;

        // Resolve custom application config
        const config = (this.config = await gatherConfig(hostElement, options));
        this.checkAbort();
        LOG.debug("Application config is", config);

        // Decide on locale and load i18n messages (if any).
        const i18n = await initI18n(options.appMetadata, config.locale);
        this.checkAbort();

        // Setup application root node in the shadow dom
        const container = (this.container = createContainer(i18n.locale));
        const styles = this.initStyles();
        shadowRoot.replaceChildren(container, styles);

        // Launch the service layer
        const { serviceLayer, packages } = this.initServiceLayer({
            container,
            properties: config.properties,
            i18n
        });
        await this.initAPI(serviceLayer);
        this.checkAbort();

        // Launch react
        this.reactIntegration = new ReactIntegration({
            rootNode: container,
            container: shadowRoot,
            serviceLayer,
            packages
        });
        this.render();
        this.state = "started";

        LOG.debug("Application started");
    }

    private render() {
        this.reactIntegration?.render(this.options.component ?? emptyComponent, {});
    }

    private initStyles() {
        const styles = this.options.appMetadata?.styles;
        const styleNode = document.createElement("style");
        applyStyles(styleNode, styles);
        if (import.meta.hot) {
            this.stylesWatch = styles?.on?.("changed", () => {
                LOG.debug("Application styles changed");
                applyStyles(styleNode, styles);
            });
        }
        return styleNode;
    }

    private initServiceLayer(config: {
        container: HTMLDivElement;
        properties: ApplicationProperties;
        i18n: AppI18n;
    }) {
        const options = this.options;
        const { container, properties, i18n } = config;
        const packageMetadata = options.appMetadata?.packages ?? {};
        const builtinPackage = createBuiltinPackage({
            host: this.hostElement,
            shadowRoot: this.shadowRoot,
            container: container,
            locale: i18n.locale,
            supportedLocales: i18n.supportedLocales
        });
        const { serviceLayer, packages } = createServiceLayer({
            packageMetadata,
            builtinPackage,
            properties,
            i18n
        });
        this.serviceLayer = serviceLayer;

        if (LOG.isDebug()) {
            LOG.debug("Launching service layer with packages", Object.fromEntries(packages));
        }
        serviceLayer.start();
        return { serviceLayer, packages };
    }

    private async initAPI(serviceLayer: ServiceLayer) {
        const result = serviceLayer.getService(
            "@open-pioneer/runtime",
            {
                interfaceName: RUNTIME_API_SERVICE
            },
            { ignoreDeclarationCheck: true }
        );
        if (result.type !== "found") {
            throw new Error(
                ErrorId.INTERNAL,
                `Failed to find instance of 'runtime.ApiService' (result type '${result.type}').` +
                    ` This is a builtin service that must be present exactly once.`
            );
        }

        const apiService = result.value.getInstanceOrThrow() as ApiService;
        try {
            const api = (this.api = await apiService.getApi());
            LOG.debug("Application API initialized to", api);
            this.apiPromise?.resolve(api);
        } catch (e) {
            throw new Error(ErrorId.INTERNAL, "Failed to gather the application's API methods.", {
                cause: e
            });
        }
    }

    private checkAbort() {
        if (this.state === "destroyed") {
            throwAbortError();
        }
    }
}

function createContainer(locale: string) {
    // Setup application root node in the shadow dom
    const container = document.createElement("div");
    container.classList.add("pioneer-root");
    container.style.minHeight = "100%";
    container.style.height = "100%";
    if (locale) {
        container.lang = locale;
    }
    return container;
}

function createServiceLayer(config: {
    packageMetadata: Record<string, PackageMetadata> | undefined;
    properties: ApplicationProperties;
    builtinPackage: PackageRepr;
    i18n: AppI18n;
}) {
    const { packageMetadata, properties, builtinPackage, i18n } = config;

    let packages: PackageRepr[];
    try {
        packages = createPackages(packageMetadata ?? {}, i18n, properties);
    } catch (e) {
        throw new Error(ErrorId.INVALID_METADATA, "Failed to parse package metadata.", {
            cause: e
        });
    }

    // Add builtin services defined within this package.
    {
        const index = packages.findIndex((pkg) => pkg.name === builtinPackage.name);
        if (index >= 0) {
            packages.splice(index, 1);
        }
        packages.push(builtinPackage);
    }

    const forcedReferences: ReferenceSpec[] = [
        {
            interfaceName: RUNTIME_API_SERVICE
        }
    ];
    const serviceLayer = new ServiceLayer(packages, forcedReferences);
    return {
        packages: new Map(packages.map((pkg) => [pkg.name, pkg])),
        serviceLayer: serviceLayer
    };
}

/**
 * Gathers application properties by reading them from the options object
 * and by (optionally) invoking the `resolveProperties` hook.
 */
async function gatherConfig(hostElement: HTMLElement, options: CustomElementOptions) {
    let configs: ApplicationConfig[];
    try {
        configs = [
            options.config ?? {},
            (await options.resolveConfig?.({
                getAttribute(name) {
                    return hostElement.getAttribute(name) ?? undefined;
                }
            })) ?? {}
        ];
    } catch (e) {
        throw new Error(
            ErrorId.CONFIG_RESOLUTION_FAILED,
            "Failed to resolve application properties.",
            {
                cause: e
            }
        );
    }

    return mergeConfigs(configs);
}

/**
 * Merges application configurations into a single object.
 * Properties / config parameters at a later position overwrite properties from earlier ones.
 */
function mergeConfigs(configs: ApplicationConfig[]): Required<ApplicationConfig> {
    // Merge simple values by assigning them in order
    const mergedConfig: Required<ApplicationConfig> = Object.assign(
        {
            locale: undefined,
            properties: {}
        } satisfies ApplicationConfig,
        ...configs
    );

    // Deep merge for application properties
    const mergedProperties: ApplicationProperties = (mergedConfig.properties = {});
    for (const config of configs) {
        for (const [packageName, packageProperties] of Object.entries(config.properties ?? {})) {
            const mergedPackageProps = (mergedProperties[packageName] ??= {});
            Object.assign(mergedPackageProps, packageProperties);
        }
    }

    return mergedConfig;
}

const DISABLE_INHERIT = ":host { all: initial; display: block; }";

// Applies application styles to the given style node.
// Can be called multiple times in development mode to implement hot reloading.
function applyStyles(styleNode: HTMLStyleElement, styles: ObservableBox<string> | undefined) {
    const cssValue = styles?.value ?? "";
    const cssNode = document.createTextNode([DISABLE_INHERIT, cssValue].join("\n"));
    styleNode.replaceChildren(cssNode);
}

function logError(e: unknown) {
    if (e instanceof Error) {
        const chain = getErrorChain(e).reverse();
        if (chain.length === 1) {
            console.error(e);
            return;
        }

        let n = 1;
        for (const error of chain) {
            console.error(`#${n}`, error);
            ++n;
        }
    } else {
        console.error("Unexpected error", e);
    }
}

function emptyComponent() {
    return null;
}
