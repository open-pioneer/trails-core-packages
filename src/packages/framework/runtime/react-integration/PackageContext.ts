// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { createContext } from "react";
import { PackageIntl } from "../i18n";
import { Service } from "../Service";
import { UseServiceOptions } from "./hooks";

export interface PackageContextMethods {
    getService: (packageName: string, interfaceName: string, options: UseServiceOptions) => Service;
    getServices: (packageName: string, interfaceName: string) => Service[];
    getProperties: (packageName: string) => Readonly<Record<string, unknown>>;
    getIntl(packageName: string): PackageIntl;
}

export const PackageContext = createContext<PackageContextMethods | null>(null);
