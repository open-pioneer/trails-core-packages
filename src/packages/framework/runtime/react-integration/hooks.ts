// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { useContext, useMemo } from "react";
import { Error } from "@open-pioneer/core";
import { InterfaceName, ServiceType } from "../ServiceRegistry";
import { PackageContext as PackageContext, PackageContextMethods } from "./PackageContext";
import { ErrorId } from "../errors";
import { PackageIntl } from "../i18n";

export interface UseServiceOptions {
    /** An additional qualifier to disambiguate service instances that implement the same interface. */
    qualifier?: string;
}

/**
 * Returns a service that implements the given interface.
 * Receives the package name of the importing package as a string.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useServiceInternal<Interface extends InterfaceName>(
    packageName: string,
    interfaceName: Interface,
    options?: UseServiceOptions
): ServiceType<Interface>;
export function useServiceInternal(
    packageName: string,
    interfaceName: string,
    options?: UseServiceOptions
): unknown;
export function useServiceInternal(
    packageName: string,
    interfaceName: string,
    options?: UseServiceOptions
): unknown {
    const context = useContext(PackageContext);
    const service = useMemo(
        () =>
            checkContext("useService", context).getService(
                packageName,
                interfaceName,
                options ?? {}
            ),
        [context, packageName, interfaceName, options]
    );
    return service;
}

/**
 * Returns all services that implement the given interface.
 * Receives the package name of the importing package as a string.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useServicesInternal<Interface extends InterfaceName>(
    packageName: string,
    interfaceName: Interface
): ServiceType<Interface>[];
export function useServicesInternal(packageName: string, interfaceName: string): unknown[];
export function useServicesInternal(packageName: string, interfaceName: string): unknown[] {
    const context = useContext(PackageContext);
    const services = useMemo(
        () => checkContext("useServices", context).getServices(packageName, interfaceName),
        [context, packageName, interfaceName]
    );
    return services;
}

/**
 * Returns the properties for the given package.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function usePropertiesInternal(packageName: string): Readonly<Record<string, unknown>> {
    const context = useContext(PackageContext);
    return checkContext("useProperties", context).getProperties(packageName);
}

/**
 * Returns the i18n object for the given package.
 *
 * This is an internal hook that is typically called indirectly via the hook
 * provided from `"open-pioneer:react-hooks"`.
 *
 * @private
 */
export function useIntlInternal(packageName: string): PackageIntl {
    const context = useContext(PackageContext);
    return checkContext("useIntl", context).getIntl(packageName);
}

function checkContext(
    hookName: string,
    contextData: PackageContextMethods | null
): PackageContextMethods {
    if (!contextData) {
        throw new Error(
            ErrorId.INTERNAL,
            `"Failed to access package context from '${hookName}': react integration was not set up properly.`
        );
    }
    return contextData;
}
