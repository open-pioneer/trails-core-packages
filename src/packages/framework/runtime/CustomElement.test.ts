/**
 * @vitest-environment jsdom
 */
import { isAbortError } from "@open-pioneer/core";
import {
    defineComponent,
    renderComponent,
    renderComponentShadowDOM
} from "@open-pioneer/test-utils/web-components";
import { waitFor } from "@testing-library/dom";
import { createElement } from "react";
import { expect, it, describe, vi, afterEach } from "vitest";
import { ApiExtension, ApiMethods, ApplicationContext } from "./api";
import { ApplicationElement, createCustomElement, CustomElementOptions } from "./CustomElement";
import { createBox } from "./metadata";
import { usePropertiesInternal } from "./react-integration";
import { ServiceOptions } from "./Service";
import { expectAsyncError } from "./test-utils/expectError";

describe("simple rendering", function () {
    const SIMPLE_STYLE = ".test { color: red }";
    const SIMPLE_ELEM = createCustomElement({
        component: () => createElement("div", { className: "test" }, "hello world"),
        appMetadata: {
            styles: createBox(SIMPLE_STYLE)
        }
    });
    customElements.define("simple-elem", SIMPLE_ELEM);

    it("should return html element", () => {
        expect(new SIMPLE_ELEM()).toBeInstanceOf(HTMLElement);
    });

    it("should render html", async () => {
        const { queries } = await renderComponentShadowDOM("simple-elem");
        const div = await queries.findByText("hello world");
        expect(div.className).toBe("test");
    });

    it("should render use styles", async () => {
        const { shadowRoot } = await renderComponentShadowDOM("simple-elem");
        const style = shadowRoot.querySelector("style")!;
        expect(style).toMatchSnapshot();
    });

    it("should clean up its content when removed from the dom", async () => {
        const { node, shadowRoot, innerContainer } = await renderComponentShadowDOM("simple-elem");
        node.remove();

        // Wait until divs are gone
        await waitFor(() => {
            const div = innerContainer.querySelector("div");
            if (div) {
                throw new Error("content still not destroyed");
            }
        });

        expect(shadowRoot.innerHTML).toBe("");
    });
});

it("explicitly setting the shadow dom mode hides the shadow root", async () => {
    function TestComponent() {
        return createElement("span", undefined, "Hello World");
    }

    const elem = createCustomElement({
        component: TestComponent,
        openShadowRoot: false
    });
    const { node } = await renderComponent(elem);
    expect(node.shadowRoot).toBeNull();
});

it("allows customization of package properties", async () => {
    const elem = createCustomElement({
        component: function TestComponent() {
            const properties = usePropertiesInternal("test");
            return createElement("span", undefined, properties.greeting as string);
        },
        appMetadata: {
            packages: {
                test: {
                    name: "test",
                    properties: {
                        greeting: {
                            value: "Hello World"
                        }
                    }
                }
            }
        },
        config: {
            properties: {
                test: {
                    greeting: "Hello User"
                }
            }
        }
    });

    const { queries } = await renderComponentShadowDOM(elem);
    const span = await queries.findByText("Hello User");
    expect(span.tagName).toBe("SPAN");
});

it("allows customization of package properties through a callback", async () => {
    const elem = createCustomElement({
        component: function TestComponent() {
            const properties = usePropertiesInternal("test");
            return createElement("span", undefined, properties.greeting as string);
        },
        appMetadata: {
            packages: {
                test: {
                    name: "test",
                    properties: {
                        greeting: {
                            value: "Hello World"
                        }
                    }
                }
            }
        },
        config: {
            properties: {
                test: {
                    greeting: "Hello User"
                }
            }
        },
        // properties from this callback take precedence
        async resolveConfig() {
            return {
                properties: {
                    test: {
                        greeting: "Bye User"
                    }
                }
            };
        }
    });

    const { queries } = await renderComponentShadowDOM(elem);
    const span = await queries.findByText("Bye User");
    expect(span.tagName).toBe("SPAN");
});

it("provides access to html containers through the application context", async () => {
    let hostElement: HTMLElement | undefined;
    let shadowRoot: ShadowRoot | undefined;
    let container: HTMLElement | undefined;

    class TestService {
        constructor(options: ServiceOptions<{ ctx: ApplicationContext }>) {
            const ctx = options.references.ctx;
            hostElement = ctx.getHostElement();
            shadowRoot = ctx.getShadowRoot();
            container = ctx.getApplicationContainer();
        }
    }

    const elem = createCustomElement({
        appMetadata: {
            packages: {
                test: {
                    name: "test",
                    services: {
                        testService: {
                            name: "testService",
                            clazz: TestService,
                            references: {
                                ctx: {
                                    name: "runtime.ApplicationContext"
                                }
                            },
                            provides: [
                                {
                                    name: "testInterface"
                                }
                            ]
                        }
                    },
                    ui: {
                        references: [
                            // Hack to start the service
                            {
                                name: "testInterface"
                            }
                        ]
                    }
                }
            }
        }
    });
    const {
        node: actualHostElement,
        shadowRoot: actualShadowRoot,
        innerContainer: actualContainer
    } = await renderComponentShadowDOM(elem);

    expect(hostElement).toBe(actualHostElement);
    expect(shadowRoot).toBe(actualShadowRoot);
    expect(container).toBe(actualContainer);
});

describe("element API", () => {
    it("should throw an error when trying to use the element's API without mounting it first", async () => {
        const elem = createCustomElement({});
        const tag = defineComponent(elem);
        const node = document.createElement(tag) as ApplicationElement;
        const error = await expectAsyncError(() => node.when());
        expect(error).toMatchSnapshot();
    });

    it("should provide an empty API by default", async () => {
        const elem = createCustomElement({});
        const { node } = await renderComponent(elem);
        const api = await (node as ApplicationElement).when();
        expect(api).toEqual({});
    });

    it("throws an error when the component is unmounted while waiting for the API", async function () {
        const elem = createCustomElement({});
        const tag = defineComponent(elem);
        const node = document.createElement(tag) as ApplicationElement;
        document.body.appendChild(node);

        const err = expectAsyncError(() => node.when());
        node.remove();

        const error = await err;
        expect(isAbortError(error)).toBe(true);
    });

    it("should allow services to provide an API", async () => {
        const events: string[] = [];
        class Extension implements ApiExtension {
            async getApiMethods(): Promise<ApiMethods> {
                return {
                    add(x: number, y: number) {
                        events.push("add");
                        return x + y;
                    },
                    otherMethod() {
                        events.push("otherMethod");
                    }
                };
            }
        }

        const elem = createCustomElement({
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            testService: {
                                name: "testService",
                                provides: [
                                    {
                                        name: "integration.ApiExtension"
                                    }
                                ],
                                clazz: Extension
                            }
                        }
                    }
                }
            }
        });

        const { node } = await renderComponent(elem);
        const api = await (node as ApplicationElement).when();
        expect(Object.keys(api).sort()).toEqual(["add", "otherMethod"]);
        expect(events).toEqual([]);

        const sum = api.add!(3, 4);
        expect(sum).toEqual(7);
        expect(events).toEqual(["add"]);

        api.otherMethod!();
        expect(events).toEqual(["add", "otherMethod"]);
    });
});

describe("i18n support", function () {
    afterEach(() => {
        vi.restoreAllMocks();
    });

    it("chooses locale from the supported browser locales (german)", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["de-DE", "de", "en"]);

        const { locale, message } = await launchApp();
        expect(locale).toBe("de-DE");
        expect(message).toBe("Hallo Welt");
    });

    it("chooses locale from the supported browser locales (english)", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["en-US", "en"]);

        const { locale, message } = await launchApp();
        expect(locale).toBe("en-US");
        expect(message).toBe("Hello world");
    });

    it("supports forcing to a specific locale", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["en-US", "en"]);

        const { locale, message } = await launchApp({
            config: {
                locale: "de-simple"
            }
        });
        expect(locale).toBe("de-simple");
        expect(message).toBe("Hallo Welt (einfach)");
    });

    it("supports forcing to a specific locale with resolveConfig()", async () => {
        const spy = vi.spyOn(window.navigator, "languages", "get");
        spy.mockReturnValue(["en-US", "en"]);

        const { locale, message } = await launchApp({
            config: {
                locale: "en" // overwritten by resolveConfig
            },
            resolveConfig() {
                return Promise.resolve({
                    locale: "de-simple"
                });
            }
        });
        expect(locale).toBe("de-simple");
        expect(message).toBe("Hallo Welt (einfach)");
    });

    /**
     * Runs an app with mocked services and i18n and returns the inner locale + translated message.
     */
    async function launchApp(
        options?: Partial<CustomElementOptions>
    ): Promise<{ locale: string; message: string }> {
        class TestService implements ApiExtension {
            private locale: string;
            private message: string;

            constructor(options: ServiceOptions<{ ctx: ApplicationContext }>) {
                const ctx = options.references.ctx;
                this.locale = ctx.getLocale();
                this.message = options.intl.formatMessage({ id: "greeting" });
            }

            async getApiMethods(): Promise<ApiMethods> {
                return {
                    getValues: () => {
                        return {
                            locale: this.locale,
                            message: this.message
                        };
                    }
                };
            }
        }

        const elem = createCustomElement({
            ...options,
            appMetadata: {
                packages: {
                    test: {
                        name: "test",
                        services: {
                            testService: {
                                name: "testService",
                                clazz: TestService,
                                references: {
                                    ctx: {
                                        name: "runtime.ApplicationContext"
                                    }
                                },
                                provides: [
                                    {
                                        name: "integration.ApiExtension"
                                    }
                                ]
                            }
                        }
                    }
                },
                locales: ["de", "en", "de-simple"],
                async loadMessages(locale) {
                    switch (locale) {
                        case "en":
                            return {
                                test: {
                                    greeting: "Hello world"
                                }
                            };
                        case "de":
                            return {
                                test: {
                                    greeting: "Hallo Welt"
                                }
                            };
                        case "de-simple":
                            return {
                                test: {
                                    greeting: "Hallo Welt (einfach)"
                                }
                            };
                    }
                    throw new Error("Unsupported locale: " + locale);
                }
            }
        });

        const { node } = await renderComponentShadowDOM(elem);
        const api = await (node as ApplicationElement).when();
        return api.getValues!();
    }
});
