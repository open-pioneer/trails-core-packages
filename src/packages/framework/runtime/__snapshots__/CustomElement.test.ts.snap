// Vitest Snapshot v1

exports[`element API > should throw an error when trying to use the element's API without mounting it first 1`] = `[Error: runtime:element-not-mounted: Cannot use the application's API because the HTML element has not yet been mounted into the DOM.]`;

exports[`simple rendering > should render use styles 1`] = `
<style>
  :host { all: initial }
.test { color: red }
</style>
`;
